File: samples/extensions/TIBCO/activity/concat/activity.ts
/*
 * Copyright © 2017. TIBCO Software Inc.
 * This file is subject to the license terms contained
 * in the license file that is distributed with this file.
 */
import {Observable} from "rxjs/Observable";
import {Injectable, Injector, Inject} from "@angular/core";
import {Http} from "@angular/http";
import {
    WiContrib,
    WiServiceHandlerContribution,
    IValidationResult,
    ValidationResult,
    IFieldDefinition,
    IActivityContribution
} from "wi-studio/app/contrib/wi-contrib";

@WiContrib({})
@Injectable()
export class ConcatActivityContributionHandler extends WiServiceHandlerContribution {
    constructor( @Inject(Injector) injector) {
        super(injector);
    }

    value = (fieldName: string, context: IActivityContribution): Observable<any> | any => {
        if(fieldName === "separator") {
           let list: Array<string> = ["-", "$", "#"];
           return list;
        } 
        return null;
    }
 
    validate = (fieldName: string, context: IActivityContribution): Observable<IValidationResult> | IValidationResult => {
       if (fieldName === "separator") {
         let vresult: IValidationResult = ValidationResult.newValidationResult();
         let useSeparatorFieldDef: IFieldDefinition = context.getField("useSeparator"); 
         let separatorFieldDef: IFieldDefinition = context.getField("separator");
         if (useSeparatorFieldDef.value && useSeparatorFieldDef.value === true) {
             if (separatorFieldDef.display && separatorFieldDef.display.visible == false) {
                 vresult.setVisible(true);
             } 
             if (separatorFieldDef.value === null || separatorFieldDef.value === "") {
               vresult.setError("TIBCO-CONCAT-1000", "Separator must be configured");
             } 
         } else {
            vresult.setVisible(false);
         }
         return vresult;
       }
      return null; 
    }
}

File: samples/extensions/TIBCO/activity/concat/activity.go
/*
 * Copyright © 2020. TIBCO Software Inc.
 * This file is subject to the license terms contained
 * in the license file that is distributed with this file.
 */
package concat

import (
	"github.com/project-flogo/core/activity"
	"github.com/project-flogo/core/support/log"
)

var activityLog = log.ChildLogger(log.RootLogger(), "tibco-activity-concat")

func init() {
	_ = activity.Register(&ConcatActivity{}, New)
}

var activityMd = activity.ToMetadata(&Input{}, &Output{})

func New(ctx activity.InitContext) (activity.Activity, error) {
	return &ConcatActivity{}, nil
}

type ConcatActivity struct {
}

func (a *ConcatActivity) Metadata() *activity.Metadata {
	return activityMd
}
func (a *ConcatActivity) Eval(context activity.Context) (done bool, err error) {
	activityLog.Info("Executing Concat activity")

	input := &Input{}
	context.GetInputObject(input)

	//Read Inputs
	if len(input.FirstStr) <= 0 {
		// First string is not configured
		// return error to the engine
		return false, activity.NewError("First string is not configured", "CONCAT-4001", nil)
	}

	if len(input.SecondStr) <= 0 {
		// Second string is not configured
		// return error to the engine
		return false, activity.NewError("Second string is not configured", "CONCAT-4002", nil)
	}

	if input.UseSeparator && len(input.Separator) <= 0 {
		// Separator is not configured
		// return error to the engine
		return false, activity.NewError("Separator is not configured", "CONCAT-4003", nil)
	}

	output := &Output{}
	if input.UseSeparator {
		//Use separator in concatenation
		output.Result = input.FirstStr + input.Separator + input.SecondStr
	} else {
		//No separator in concatenation
		output.Result = input.FirstStr + input.SecondStr
	}

	err = context.SetOutputObject(output)
	if err != nil {
		return false, err
	}
	return true, nil
}
